kratos/modeler/key_plane_generation/key_plane_generation_with_refinement.cpp

```diff
@@ -0,0 +1,251 @@
+    array_1d<double,3> min_v(3,std::numeric_limits<double>::max()), max_v(3, -std::numeric_limits<double>::max());
+    // Find Bounding box
+    for(auto& r_node : rMyModelPart.Nodes()){
+        double x = r_node.X(), y = r_node.Y(), z = r_node.Z();
+        min_v[0] = std::min(min_v[0], x); max_v[0] = std::max(max_v[0], x);
+        min_v[1] = std::min(min_v[1], y); max_v[1] = std::max(max_v[1], y);
+        min_v[2] = std::min(min_v[2], z); max_v[2] = std::max(max_v[2], z);
+    }
+    return std::make_pair(min_v, max_v);
+}
+bool KeyPlaneGenerationWithRefinement::IsEmpty(Parameters param)
+{
+    return param.WriteJsonString() == "{}";
+}
+
+void KeyPlaneGenerationWithRefinement::FindRefinements(std::vector<BoundingBoxType>& rRefinementAreas, std::vector<array_1d<double,3>>& rVoxelSizes,const array_1d<double,3>& rGlobalVoxelSize){
+    Parameters parameters = this->GetParameters();
+    for(auto refinement_parameters : parameters["refinement_zones"]){
+        // Now we either select the given refinement zone or compute it using a modelpart
+        BoundingBoxType refinement_zone;
+        if(refinement_parameters["refined_modelpart"].GetString() != ""){ // ModelParts are not empty and we are going to find the bounding box
+            const ModelPart& r_refined_modelpart = GetModelPart(refinement_parameters["refined_modelpart"].GetString());
+            if(r_refined_modelpart.NumberOfNodes() > 0) {
+                refinement_zone = GetBoundingBox(r_refined_modelpart);
+                rRefinementAreas.push_back(refinement_zone);
+            }
+        } else {
+            refinement_zone.first = refinement_parameters["refined_zone"]["min_point"].GetVector();
+            refinement_zone.second = refinement_parameters["refined_zone"]["max_point"].GetVector();
+            rRefinementAreas.push_back(refinement_zone);
+            //TODO avoid the refinament area to be defined outside the global bounding box
+        }
+        // Now we add the element size on each direction
+        array_1d<double,3> dx_local = refinement_parameters["voxel_sizes_ratio"].GetVector();
+        dx_local[0] *= rGlobalVoxelSize[0]; dx_local[1] *= rGlobalVoxelSize[1]; dx_local[2] *= rGlobalVoxelSize[2];
+        rVoxelSizes.push_back(dx_local);
+    }
+}
+
+std::pair<array_1d<std::vector<double>,3>,array_1d<std::vector<double>,3>> KeyPlaneGenerationWithRefinement::ComputePartitionsAndVoxelSizeByDirection(std::vector<BoundingBoxType>& rRefinementAreas,
+                                                std::vector<array_1d<double,3>>& rVoxelSizes,
+                                                BoundingBoxType& rGlobalBoundingBox,
+                                                const array_1d<double,3>& rGlobalVoxelSize){
+    array_1d<std::vector<double>,3> partitions;
+    array_1d<std::vector<double>,3> partitions_voxel_sizes;
+    for(int i_direction = 0 ; i_direction < 3 ; i_direction++){
```

Comment:
`std::size_t`. Happens on many other loops, to keep the review clean I will not add more.